cwl/src/main/scala/cwl/CommandLineTool.scala

```diff
@@ -150,38 +150,102 @@ case class CommandLineTool private(
     }
   }
 
+
+  private def environmentDefs(requirementsAndHints: List[Requirement], expressionLib: ExpressionLib): ErrorOr[Map[String, WomExpression]] = {
+    // For environment variables we need to make sure that we aren't being asked to evaluate expressions from a containing
+    // workflow step or its containing workflow or anything containing the workflow. The current structure of this code
+    // is not prepared to evaluate those expressions. Actually this is true for attributes too and we're totally not
+    // checking for this condition there. Blurgh.
+    // TODO CWL: for runtime attributes, detect unevaluatable expressions in the containment hierarchy.
+
+    // This traverses all `EnvironmentDef`s within all `EnvVarRequirement`s. The spec doesn't appear to say how to handle
+    // duplicate `envName` keys in a single array of `EnvironmentDef`s; this code gives precedence to the last occurrence.
+    val allEnvVarDefs = for {
+      req <- requirementsAndHints
+      envVarReq <- req.select[EnvVarRequirement].toList
+      // Reverse the defs within an env var requirement so that when we fold from the right below the later defs
+      // will take precedence over the earlier defs.
```

Comment:
Also does not `reverse` -> `foldRight` == `foldLeft` ?

Reply:
@danbills well this has to do something, I only documented how what I chose to do works. 🙂 

@Horneth it would except that the containing `EnvVarRequirement`s are iterated in the opposite order. I have to reverse one or the other.